include/slb/detail/invoke.hpp

```diff
@@ -0,0 +1,384 @@
+/*
+  SLB.Detail.Invoke
+
+  Copyright Michael Park, 2017
+  Copyright Agustin Berge, 2017
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+*/
+
+#ifndef SLB_DETAIL_INVOKE_HPP
+#define SLB_DETAIL_INVOKE_HPP
+
+#include "config.hpp"
+#include "lib.hpp"
+
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace slb {
+namespace detail {
+namespace _invoke {
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename T>
+struct is_reference_wrapper : std::false_type {};
+
+template <typename T>
+struct is_reference_wrapper<std::reference_wrapper<T>> : std::true_type {};
+
+////////////////////////////////////////////////////////////////////////////////
+struct ref_tag {};
+struct ref_wrapper_tag {};
+struct ptr_tag {};
+
+template <typename C, typename T1, typename Enable = void>
+struct dispatch_mem_ptr;
```

Comment:
Why not just `using type = ptr_tag;` here? I guess it's a more general style question of:
do we want a forward declaration followed by specializations and the generic definition, or
do we just want the generic definition, followed by specializations. I don't have a strong
preference but I think we should pick one, unless you think it should be situational.

Reply:
In this particular instance, because the specification is positional: the base template depends on "the previous two items".

In general, I don't expect the difference to matter. I think I would intuitively go with base template first then specializations when the specification carves out exceptions, and with ordered definitions when the specification consist of sequential steps.

Reply:
Oh I see what you're saying. It even shows up in comment form:
```
// `ptr_tag` when `pm` is a pointer to member of a class `C` and `t1` does not
// satisfy the previous two items;
```
If we were to define it in the generic version we'd have to refer to "the following two items",
which would deviate from the standard text.